File: Army(java)/SpellCaster.java
import java.util.HashMap;
import java.util.Map;

public abstract class SpellCaster extends Soldier {
    protected Mana mana = new Mana();
    protected enum SpellList { FIREBALL, FREEZ, FLOOD, AVALANCHE, PATCHUP, HEAL }
    protected Map<SpellList, Spell> spellBook = new HashMap<>();

    {
        this.getHealth().setHitPointsLimit(50); //initializationBlock copies in every constructor
        this.getHealth().setHitPoints(50);
        this.getAttack().setDamage(3);
    }

    public SpellCaster(String name) {
        super(name);
    }

    public Mana getMana() {
        return this.mana;
    }

    public abstract void spellCast(Unit enemy);

    protected class Mana {
        private int manaLimit = 100;
        private int currentMana = manaLimit;

        public void takeMana(int mana) throws OutOfManaException {
            if ( this.currentMana < mana ) {
                throw new OutOfManaException();
            }
            this.currentMana -= mana;
        }

        public int getManaLimit() {
            return this.manaLimit;
        }
        public void setManaLimit(int manaLimit) {
            this.manaLimit = manaLimit;
        }

        public int getCurrentMana() {
            return this.currentMana;
        }
        public void setCurrentMana(int currentMana) {
            this.currentMana = currentMana;
        }
    }

    @Override
    public String toString() {
        return super.toString() + " mana: " + this.getMana().getCurrentMana() + "/" + this.getMana().getManaLimit();
    }
}

File: Army(kotlin)/SpellCaster.kt
abstract class SpellCaster(name: String): Soldier(name) {
    val mana: Mana = Mana()
    protected enum class SpellList {
        FIREBALL,
        FREEZ,
        FLOOD,
        AVALANCHE,
        PATCHUP,
        HEAL,
    }
    protected val spellBook = mutableMapOf<SpellList, Spell>()

    init {
        this.health.hitPointsLimit = 50
        this.health.hitPoints = 50
        this.attack.damage = 3
    }

    abstract fun spellCast(enemy: Unit)

    inner class Mana {
        var manaLimit = 100
        var currentMana = manaLimit

        fun takeMana(mana: Int) {
//            if ( this.currentMana < mana ) throw OutOfManaException()
            check(this.currentMana >= mana) { "Out of mana" }
            this.currentMana -= mana
        }
    }

    override fun toString(): String {
        return super.toString() + " mana: ${this.mana.currentMana}/${this.mana.manaLimit}"
    }
}
//class OutOfManaException: Exception("OutOfManaException") {}